Fragment Caching for Renderables

Matt-Yorkley opened the pull request:
### Summary

Hello Rails core :wave: I'm trying to explore options for getting Fragment Caching to work with Renderables. I'll start at the beginning...

#### WTF is a Renderable?

A [ViewComponent](https://github.com/github/view_component) class is a prime example on the fully-featured end of the scale, but a "Renderable" can also be a humble PORO which encapsulates some custom rendering logic and implements `#render_in` or `#to_partial_path`, as per [this example.](https://gist.github.com/websebdev/b9a2fc018337acdb45faae04cec0ac47)

Ultimately their usage looks a bit like this:

```ruby
<%= render MessageRenderable.new("Hello world!") %>
```

#### Renderables in Rails

Currently, `ActionView` has various bits of explicit support for using Renderables in the view layer, for example here in [ActionView::Renderer](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionview/lib/action_view/rendering.rb#L147-L148) or in [ActionView::Template::Renderable](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionview/lib/action_view/template/renderable.rb).

I'm interested in bringing support for Renderables into the parts of `ActionView` responsible for **Fragment Caching**, as there is currently no compatibility at all, meaning Fragment Caching is unusable.

I'm trying to approach this in a way that treats Renderables as a relatively generic duck-type that isn't tied to any specific implementation, so that `ActionView` itself doesn't need to know the details of what or how any given Renderable class is doing under the hood.

#### Adding Fragment Caching Support

The missing piece of the puzzle revolves around the creation of digests and mapping of dependencies in `ActionView::Digestor` and `ActionView::DependencyTracker` respectively. Renderables are currently ignored, which means that any changes in related files are not reflected in the digests used to create [composite cache keys](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionview/lib/action_view/helpers/cache_helper.rb#L27-L28) used for fragments, so the cache entries are not invalidated when they should be (when the files' contents change).

Adding this support requires some minor changes so that;
- when render dependencies are being enumerated by `ActionView::DependencyTracker` it recognises render calls involving Renderables
- when `ActionView::Digestor` receives a list of dependencies that includes Renderables, it handles them slightly differently to regular partials

Matt-Yorkley committed:
Remove unnecessary keys argument from Digestor#find_template

This `keys` argument is always an empty array, and doesn't need to be part of the interface.

Matt-Yorkley committed:
Remove unnecessary children argument from Node initialize

This argument doesn't get passed in any time Node is initialized, the `children` attribute is only used after initialization.

Matt-Yorkley committed:
Add support for Renderables to TemplatePath

Matt-Yorkley committed:
Add support for Renderables to Digestor

Matt-Yorkley committed:
Remove logical_name from Digestor

I'm pretty sure this dead code, especially since the recent TemplatePath changes. Simplifies the class and removes some allocations.

Matt-Yorkley commented:
Ping @HParker it would be great to get your thoughts on this. FYI I've got an alternate version of this PR which integrates with [RipperTracker](https://github.com/rails/rails/pull/42458) instead of `ERBTracker`...

Matt-Yorkley committed:
Register mime type for :rb

File lookups for views use ActionView::Resolver::PathParser#build_path_regex, which ignores unregistered mime types.

HParker commented:
Cc @jhawthorn

oboxodo commented on actionview/lib/action_view/dependency_tracker.rb:
```diff
@@ -59,6 +61,12 @@ class ERBTracker # :nodoc:
 
+      # A renderable class e.g. TopicRenderable.new("foo") => "TopicRenderable"
+      RENDERABLE = /
+        (?<renderable>[A-Z][A-Za-z:]*)  # a class name captured as RENDERABLE
+        \.                              # followed by .
```
This wouldn't capture something like this:

```erb
<%- my_renderable_instance = MyRenderable.new %>
<%= render my_renderable_instance %>
```

But I guess that's too much to ask for, and I'd happily accept that limitation. But it should be documented.

Matt-Yorkley replied:
Yeah, I couldn't see any feasible way to get around that.

Matt-Yorkley committed:
Add support for Renderables as render dependencies in ERBTracker

Matt-Yorkley committed:
Use MatchData#named_captures

The MatchData object returned by Regexp#last_match throws an error if an undefined group is requested. In the case where layout dependencies are passed to #add_dependencies, the :renderable match group will not exist. MatchData#named_captures returns a simpler Hash which returns nil for undefined groups instead of raising.

Matt-Yorkley committed:
Add support for tracking dependencies of Renderables in new RipperTracker

jhawthorn commented:
Thanks for looking into this. This is very informative and well explored, but there are a few issues that would prevent us from taking this path.

This relies on the "renderable" class existing within the view path, which I don't think is the case for at least view_component, and shouldn't be expected for a generic "renderable". These classes can exist in any file/directory and are just a ruby constant, not a template. Similarly they shouldn't be represented in a `TemplatePath`.

This also assumes that processing the "renderable" `.rb` file itself will get appropriate dependencies, which isn't necessarily true. I see in the tests that the example uses a custom tracker to override this to inject the "renderable"'s custom logic, but this output is still used by dependency tracker. However this output is still going to be processed as a template path (with view path and details based lookups) which is not necessarily the case for a renderable (it isn't the case for view_component).

IMO a solution for this would need to be less tied to existing template lookup logic since that isn't how a renderable behaves.

I also wonder if this could be solved (or at least prototyped) without changes to Rails by (in the renderable railtie or similar) registering a custom tracker for `.erb`, which called the original tracker and appended dependencies for renderable renders via regex (or ripper parsing).

Matt-Yorkley commented:
Thanks for taking a look @jhawthorn :pray: I realise the PR description is maybe a bit thin on some of the finer details, so I'll put some more thoughts down here.

> This relies on the "renderable" class existing within the view path, which I don't think is the case for at least view_component, and shouldn't be expected for a generic "renderable".

Yes, renderables would need to be in the view path generally, or in the more specific case of ViewComponent, the component directory can be added to the view paths in an initializer with:
```ruby
ActionController::Base.view_paths.unshift(ViewComponent::Base.view_component_path)
```
this would be needed anyway :point_up:, as the `/components` path is full of non-component template files (erb / haml etc) which would also need to be in a discoverable location for dependency tracking. As a side note, that [ViewComponent::Base.view_component_path](https://github.com/github/view_component/pull/987) accessor exists for this purpose.

> This also assumes that processing the "renderable" .rb file itself will get appropriate dependencies, which isn't necessarily true. I see in the tests that the example uses a custom tracker to override this to inject the "renderable"'s custom logic, but this output is still used by dependency tracker.

Yeah, I added that in the tests as an illustration that some implementations might extend the base tracker(s). Any implementation using renderables may want to do that, but it would almost certainly contain logic that's internal to that implementation (and potentially call methods that are outside of Rails). So a renderable could (for example) inherit from a _parent_ class, or it could reference some other template files in a specific way, and if any of those files change the digest would need to change too. Here's a working example of a ViewComponent-specific tracker which solves both of those issues: https://gist.github.com/Matt-Yorkley/111d2f44ff29c4f12a7340aea2d8df65. It supplies additional dependencies for ViewComponent files, but it also depends on the changes in this PR.

> However this output is still going to be processed as a template path (with view path and details based lookups) which is not necessarily the case for a renderable (it isn't the case for view_component).

If you imagine `Template A` and it renders `Partial A`, `Partial B`, and `Renderable A`, and `Partial B` also renders `Renderable C` and `Renderable D`, and `Renderable D` inherits from `Renderable E`, which also renders `Template B` which renders `Partial D`... and then think about what the dependency tracking would need to look like in order to build a usable digest for `Template A`, it quickly become clear that both renderables and templates (and their paths) are very closely intertwined.

If you happen to _debug_ a digest tree object with that kind of setup, it looks a bit like this (and it works): https://gist.github.com/Matt-Yorkley/9799f3ead134711693bfc429f611ac54. I could provide a working repo using ViewComponent if you're interested?

> I also wonder if this could be solved (or at least prototyped) without changes to Rails by (in the renderable railtie or similar) registering a custom tracker for .erb, which called the original tracker and appended dependencies for renderable renders via regex (or ripper parsing).

That was where I started! :sweat_smile: The changes here are the bare minimum to enable it to work whilst keeping all the implementation-specific details out of ActionView. The alternative (implementing it entirely outside of Rails) ends up being a mess of monkey-patches and duplicated/modified ActionView code that would break any time there are any upstream changes in Rails.

dkniffin commented:
I'm still very much interested in this functionality.

@jhawthorn It looks like you never replied to @Matt-Yorkley here. Any further feedback? As he mentioned, this change is required to support fragment caching in ViewComponent without it including a bunch of monkey patches (which is what's done in the current PR here: https://github.com/github/view_component/pull/980 )

erikaxel commented:
@jhawthorn Any chance this can be revived? I loved that Rails accepted changes to allow 3rd party component libraries in 2020 (https://github.com/rails/rails/pull/36388). It seems like many in the community, like Github itself, appreciates to be able organise some its views into components, even if this is not the main Rails way. And although Github does not cache its components, many others do, and it would be great if it is possible to have cache support for 3rd party component libraries without the hacks that @Matt-Yorkley mentions.

From my reading it seems like Matt has given answers to your initial questions, but I might have misunderstood something? Would you like us to explore a different direction that is acceptable to merge? Or is there anything we can do to help this along?

jhawthorn commented:
Unfortunately the feedback hasn't been addressed.

It's an unacceptable hack to add the view components to the view path, though it may work under some configurations (and not all) today it only does by accident and is unlikely to in the future.

ViewComponents, or any other renderable, are not views. They work by passing a "renderable" object to the `render` call. Proper views on the other hand pass a string to the `render` call, which finds a template with "details" (`.html.erb`, `.xml.builder`, etc). These aren't the same operation and treating the view_component constants like they are template paths is likely to cause strange behaviour.

Having the component Ruby files (`.rb`) inside a view directory is unsupported. It can likely cause strange behaviour today(what happens if you try to render its path) and future improvements to AV are likely to cause even more issues (warnings if `.rb` isn't a supported template type, if we add template eager loading this will break).

For a solution to be accepted:

* The view component directory must not be added to the `view_paths`
* It must not be tied to the view component implementation
  * The fact that there's an `.html.erb` file which happens to match the constant name is view component specific
  * The fact we're looking for `SomeConstant.new` is VC-specific

I think it's possible, but the likely best path forward is to add something pluggable in order to determine additional dependencies from a view (because the "renderable"-specific part is otherwise something we can't solve for generally).

I also think there's some question of how compatible view components are with the concept of fragment digests. Does the digest cover the VC class definition, or just the template. If it considers the class definition, what about its superclass or anything it inherits. This is outside of Rails' scope (a problem for whoever implements this in VC 😅), and I think making tradeoffs here is completely acceptable (ex. Rails doesn't consider helpers when calculating digests), but I think it's worth considering that whatever is implemented here is likely to be confusing to users. To me this is a strong point towards more of this logic needing to live inside VC (or other renderable).

dkniffin commented:
@jhawthorn All of that makes sense.

Would it work to update Rails to support a new method for generating a cache key on a renderable object? Similar to how `render` was implemented to allow rendering an object that isn't a view template, something like `cache_key` could be implemented to create the cache key for that file. Then Rails just needs to support calling that method for any renderable object, and the view component implementation handles the logic for how that method checks if the cache should be expired or not.
